haskell/test/SMT/Common/Run.hs

```diff
@@ -0,0 +1,263 @@
+                  , ((unTAddress registryDaoConsumer), OtherContractType $ OtherContract [] (toMutez 0))
+                  ]
+              , msMutez = newBal
+              }
+
+        -- Call ligo dao and run haskell model then compare the results.
+        handleCallLoop (dao, tokenContract, registryDaoConsumer) contractCalls newMs_
+
+    )
+    emulatedImplIntegrational
+
+-- | For each generated entrypoint calls, this function does 3 things:
+-- 1. Run haskell model against the call.
+-- 2. Call ligo dao with the call
+-- 3. Compare the result. If it is to be expected, loop to the next call, else throw the error.
+handleCallLoop :: (TAddress Parameter, TAddress FA2.Parameter, TAddress (MText, (Maybe MText))) -> [ModelCall] -> ModelState -> EmulatedT PureM ()
+handleCallLoop _ [] _ = pure ()
+handleCallLoop (dao, gov, viewC) (mc:mcs) ms = do
+
+  -- All values here are needed for `printResult`. See `printResult` for the usage of the values.
+  let (haskellErrors, updatedMs) = handleCallViaHaskell mc ms
+      haskellStoreE = case haskellErrors of
+        Just err -> Left err
+        Nothing -> Right (updatedMs & msFullStorage & fsStorage)
+      haskellDaoBalance = updatedMs & msMutez
+
+      govContract = updatedMs & msContracts
+        & Map.lookup (unTAddress gov)
+        & fromMaybe (error "Governance contract does not exist")
+      haskellGovStore = case govContract of SimpleFA2ContractType c -> c & sfcStorage; _ -> error "Shouldn't happen."
+      haskellGovBalance = case govContract of SimpleFA2ContractType c -> c & sfcMutez; _ -> error "Shouldn't happen."
+
+      viewContract = updatedMs & msContracts
+        & Map.lookup (unTAddress viewC)
+        & fromMaybe (error "View contract does not exist")
+      haskellViewStore = case viewContract of OtherContractType c -> c & ocStorage; _ -> error "Shouldn't happen."
+
+  (ligoStoreE, ligoDaoBalance, ligoGovStore, ligoGovBalance, ligoViewStore)
+    <- handleCallViaLigo (dao, gov, viewC) mc
+
+  printResult mc
+    (haskellStoreE, haskellDaoBalance, haskellGovStore, haskellGovBalance, haskellViewStore)
+    (ligoStoreE, ligoDaoBalance, ligoGovStore, ligoGovBalance, ligoViewStore)
+
+  handleCallLoop (dao, gov, viewC) mcs updatedMs
+
+
+-- | Compare dao's storage, error, and balance, and its auxiliary contracts:
+--   - an fa2 contract storage and mutez (which is gov contract in this case)
+--   - a view contract storage
+-- Note: Gov contract does not necessarily have to be the governance contract of basedao.
+-- We simply need a FA2 contract to do various operation in the haskell model, and gov contract
+-- just happen to be a convenience FA2 contract that we can use.
+printResult
+  :: ModelCall
+  -> (Either ModelError Storage, Mutez, [FA2.TransferParams], Mutez, [Text])
+  -> (Either ModelError Storage, Mutez, [FA2.TransferParams], Mutez, [Text])
+  -> EmulatedT PureM ()
```

Comment:
I need to use `EmulatedT` because `getStorage'` does not work in `MonadEmulated`. 

This is an old function, but I'll need to wait for this PR to be merged first: https://github.com/tqtezos/baseDAO/pull/294, to be able to use the `getStorage` that could get BigMap value.

Reply:
`getStorage'` works with `MonadEmulated` tho . You might be thinking of `MonadNettest`, in which case, yeah, `getStorage'` wouldn't work.

![image](https://user-images.githubusercontent.com/718166/125619527-9827370a-57de-4b0e-9519-711af5e8008a.png)

`EmulatedT PureM` is a specialization of `MonadEmulated`

Reply:
Nice, I didn't know about this. Yeah I was reffering to `MonadNettest`. I've updated this to use `MonadEmulated` instead now.